Improve: option exp-grouping

gpetiot opened the pull request:
Second attempt at fixing #508 & #491, the diff of test_branch looks good (it looks like `indicate-multiline-delimiters` was not taken into account for `(let ... in ...)` and now it is).

Julow commented on test/passing/exp_grouping_begin_end.ml.ref:
```diff
@@ -0,0 +1,101 @@
+  foooooooo
+    begin
+      if foooooooooooooooooooooooooooo then
+      if foooooooooooooooooooooooooooo then foooooooooooooooooo
```
This should be indented

gpetiot replied:
fixed.

gpetiot commented on src/Fmt_ast.ml:
```diff
@@ -152,6 +152,24 @@ let wrap_fun_sig_args ~stmt_loc c indent k =
+        if String.is_prefix ~prefix:"begin" str then `Begin_end else `Parens
+  in
+  match grouping_kind with
+  | `Parens -> wrap_fits_breaks_if ~space c.conf parens "(" ")" k
```
Done.

gpetiot commented:
@Julow since not all parentheses can be replaced with `begin`/`end` maybe it makes sense to only have the `parens`/`preserve` values for this option, so that we can avoid to check if it is correct to us `begin`/`end` for each expression (the code is not designed this way) ?

Julow commented:
I agree using `begin`/`end` almost everywhere is really ugly. `parens`/`preserve` are great and we could implement `begin-end` later when we have a way to decide between parenthesis and `begin`/`end` or as separate options.

gpetiot commented:
I rebased, removed the `begin-end` value, and the diff of test_branch.sh looks good.
@Khady I know some cases are not handled right now, but I think the PR can be tested on your side now, so we can improve it with your feedback.

Khady commented:
Maybe I am missing something. Because in my setup it doesn't seem to preserve the begin/end as much as in the tests of this PR.

```ocaml
let v =
  List.map begin fun x ->
    let y = x + 1 in
    let z = y + 1 in
    x + y + z
  end
    [1;2;2;3;4
    ]


let () =
  Lwt_main.run
    begin
      let a = "a" in
      let b = "b" in
      let c = "c" in
      Lwt.return "test"
    end

let () =
  List.iter begin fun v ->
    (* do a lot of things *)
    let a = "a" in
    let b = "b" in
    let c = "c" in
    ()
  end values
```

get reformated into

```ocaml
$ ocamlformat --root=. begin_end.ml --print-config
profile=ocamlformat
quiet=false
max-iters=10
comment-check=true
wrap-fun-args=true
wrap-comments=false
type-decl-indent=2
type-decl=compact
stritem-extension-indent=0
space-around-variants=false
space-around-records=false
space-around-lists=false
space-around-arrays=false
single-case=compact
sequence-style=separator
sequence-blank-line=compact
parse-docstrings=false
parens-tuple-patterns=multi-line-only
parens-tuple=always
parens-ite=false
ocp-indent-compat=false
nested-match=wrap
module-item-spacing=sparse
max-indent=none
match-indent-nested=never
match-indent=0
margin=80
let-open=preserve
let-module=compact
let-binding-spacing=compact
let-binding-indent=2
let-and=compact
leading-nested-match-parens=false
infix-precedence=indent
indicate-nested-or-patterns=space
indicate-multiline-delimiters=true
indent-after-in=0
if-then-else=compact
function-indent-nested=never
function-indent=2
field-space=tight
extension-sugar=preserve
extension-indent=2
exp-grouping=preserve (file .ocamlformat:23)
escape-strings=preserve
escape-chars=preserve
doc-comments-tag-only=default
doc-comments-padding=2
doc-comments=after
disable=false
cases-matching-exp-indent=compact
cases-exp-indent=4
break-struct=force
break-string-literals=wrap
break-sequences=false
break-separators=before
break-infix-before-func=true
break-infix=wrap
break-fun-sig=wrap
break-fun-decl=wrap
break-collection-expressions=fit-or-vertical
break-cases=nested
assignment-operator=end-line
let v =
  List.map
    (fun x ->
      let y = x + 1 in
      let z = y + 1 in
      x + y + z)
    [1; 2; 2; 3; 4]

let () =
  Lwt_main.run
    begin
      let a = "a" in
      let b = "b" in
      let c = "c" in
      Lwt.return "test"
    end

let () =
  List.iter
    (fun v ->
      (* do a lot of things *)
      let a = "a" in
      let b = "b" in
      let c = "c" in
      ())
    values
```

gpetiot commented:
With the last commits your example is formatted as:
```ocaml
let v =
  List.map
    begin
      fun x ->
      let y = x + 1 in
      let z = y + 1 in
      x + y + z
    end
    [1; 2; 2; 3; 4]

let () =
  Lwt_main.run
    begin
      let a = "a" in
      let b = "b" in
      let c = "c" in
      Lwt.return "test"
    end

let () =
  List.iter
    begin
      fun v ->
      (* do a lot of things *)
      let a = "a" in
      let b = "b" in
      let c = "c" in
      ()
    end
    values
```

Julow commented on test/passing/exp_grouping_parens.ml.ref:
```diff
@@ -0,0 +1,203 @@
+    a= b }
+
+let _ =
+  {< x = let a = b in
```
Not related to this PR but should we put parens here ? To be consistent with records.

gpetiot commented:
@Khady I would like to merge this before releasing, are there still missing cases that you need?

Khady commented:
I have a weird case, with a more complex .ocamlformat

```ocaml
let v = List.map begin fun x -> x + 1 end [1;2;3]
let v = List.map (fun x -> x + 1) [1;2;3]
```

```
ocamlformat demo.ml --print-config
profile=ocamlformat
quiet=false
max-iters=10
comment-check=true
wrap-fun-args=true (file .ocamlformat:15)
wrap-comments=false
type-decl-indent=2
type-decl=sparse (file .ocamlformat:12)
stritem-extension-indent=0
space-around-variants=true (file .ocamlformat:23)
space-around-records=true (file .ocamlformat:22)
space-around-lists=true (file .ocamlformat:21)
space-around-arrays=true (file .ocamlformat:20)
single-case=compact
sequence-style=terminator (file .ocamlformat:6)
sequence-blank-line=compact
parse-docstrings=false
parens-tuple-patterns=multi-line-only
parens-tuple=multi-line-only (file .ocamlformat:14)
parens-ite=true (file .ocamlformat:13)
ocp-indent-compat=true (file .ocamlformat:8)
nested-match=align (file .ocamlformat:24)
module-item-spacing=sparse
max-indent=2 (file .ocamlformat:25)
match-indent-nested=never
match-indent=0
margin=120 (file .ocamlformat:1)
let-open=preserve
let-module=compact
let-binding-spacing=compact
let-binding-indent=2
let-and=sparse (file .ocamlformat:3)
leading-nested-match-parens=true (file .ocamlformat:9)
infix-precedence=indent
indicate-nested-or-patterns=space
indicate-multiline-delimiters=true
indent-after-in=0
if-then-else=compact
function-indent-nested=never
function-indent=2
field-space=loose (file .ocamlformat:18)
extension-sugar=preserve
extension-indent=2
exp-grouping=preserve (file .ocamlformat:19)
escape-strings=preserve
escape-chars=preserve
doc-comments-tag-only=default
doc-comments-padding=2
doc-comments=before (file .ocamlformat:2)
disable=false
cases-matching-exp-indent=compact
cases-exp-indent=2 (file .ocamlformat:7)
break-struct=natural (file .ocamlformat:5)
break-string-literals=wrap
break-sequences=true (file .ocamlformat:17)
break-separators=after-and-docked (file .ocamlformat:11)
break-infix-before-func=false (file .ocamlformat:16)
break-infix=fit-or-vertical (file .ocamlformat:10)
break-fun-sig=wrap
break-fun-decl=wrap
break-collection-expressions=fit-or-vertical
break-cases=toplevel (file .ocamlformat:4)
assignment-operator=end-line
break-before-in=fit-or-vertical
let v = List.map begin
  fun x -> x + 1
    end [ 1; 2; 3 ]

let v = List.map (fun x -> x + 1) [ 1; 2; 3 ]
```

gpetiot commented:
Merged because despite Louis' last comment, he told me they only use begin/end for several-lines-long blocks.
